feat: completely rewrite project

xJonathanLEI opened the pull request:
This PR replaces the current code with a new implementation of the same core pipeline that's more maintainable and extensible.

Nevertheless, even the new implementation is still a work in progress. Major features missing:

- error handling
  - remove all `unwrap()` calls in background tasks
  - allow services to report fatal error and trigger an app-wide shutdown
- deeper graceful shutdown integration
- actual persistent persistence backends (currently only in-memory is supported)
- modes other than sovereign

These should be implemented/improved with follow-up PRs.

That said, the new impl is already capable of following a network and making proofs available on Celestia (the only supported DA backend at the moment) with a space-efficient binary representation. A separate demo project has been developed to follow the DA pointer to recreate the latest network state.

xJonathanLEI committed:
chore: switch to canonical `snos`

glihm commented on Cargo.toml:
```diff
@@ -1,40 +1,41 @@
+good_lp = { version = "=1.8.1", default-features = false }
+hex = { version = "0.4.3", default-features = false }
+log = "0.4.22"
+prove_block = { git = "https://github.com/keep-starknet-strange/snos", rev = "6e0514d" }
```
Will be changed for cartridge's fork once persistent is merged back, to support the first block proving and matching blockifier constants.

glihm commented on saya/core/src/prover/atlantic/mod.rs:
```diff
@@ -0,0 +1,195 @@
+impl AtlanticProver {
+    async fn run(mut self) {
+        // TODO: split this type further into sub-services to allow parallelization
+        // TODO: add persistence for in-flight proof requests to be able to resume progress
```
What would be the path for that?

Is the storage expected to only be used by the orchestrator, hence adding a new channel to notify about jobs being registered could make sense?
Like so the orchestrator can re-feed the same block in case this is necessary?

xJonathanLEI replied:
I think the best way is to localize storage to specific services that request it. The orchestrator would not know whether any of its child services use storage or not - it'd be completely internal to them.

Then for example when the prover receives a certain block number, it can check internally whether it has already submitted the request for the block from a previous run. If so, and the block content matches, it skips submission and fetches the proof from that previous run instead. This process is opaque to any external components.

xJonathanLEI committed:
Completely rewrite project

xJonathanLEI commented:
Rebased.

xJonathanLEI committed:
feat: persistent mode

glihm commented on .gitignore:
```diff
@@ -1,2 +1 @@
-/target
-blocks.db
```
This file is still being created on startup.

glihm commented on saya/core/src/data_availability/celestia.rs:
```diff
@@ -0,0 +1,183 @@
+    service::{Daemon, FinishHandle, ShutdownHandle},
+};
+
+// TODO: make namespace configurable?
```
In the current setup, there is one saya for one katana. Hence it will be important to configure it from the CLI (at least for now). Or we can update the `DataAvailabilityCursor` to contain the `chain_id` for less configuration at the high level.

In the long run, we will tend to have one saya that can aggregate jobs instead of having 1-1 matching. In this situation, a const would suffice.

glihm commented on bin/saya/programs/layout_bridge.json:
We may include compiled `snos` and `layout_bridge` to have default values using `git lfs`, wdyt?

glihm reviewed (approved):
Amazing rework here, thanks for the work done @xJonathanLEI.

Some notes:
1. Seeing several TODOs in the code. Please don't hesitate to flag using issues (or let me know and I'll create them) the one that could be achieved independently of the current WIP.
2. In a subsequent PR, SNOS will be updated to be able to prove Katana blocks, [using this branch of the fork](https://github.com/cartridge-gg/snos/tree/kariy/katana-compat-genesis).
3. Persistency in Atlantic client will be important since the two proofs generation is still relying on this service and due to the time it takes, the probability Saya is being interrupted before the layout bridge is available may be high.

As discussed, let's merge for this first iteration to have smaller PRs for the coming tasks and iterate on multiple fronts independently. 👍

xJonathanLEI committed:
feat: on-chain proof verification

xJonathanLEI committed:
feat: do not wait for Atlantic fact registration

Saya now performs on-chain fact registration interally. There's no need
to wait for Atlantic to perform fact registartion.

xJonathanLEI committed:
fix: use `relaxed` for `memory_verification`

This config results in the same fact hash expected by `piltover`.
